scheduler/scheduler.py

```diff
@@ -2668,41 +2717,41 @@ def _done_callback_extended_lease(self, job_id):
                 return
 
+            self._logger.debug('Trying to complete job {0} which had an '
+                               'extended lease...'.format(job_id))
+
+            scale_factor = self._jobs[job_id.singletons()[0]].scale_factor
             num_updates = []
             for single_job_id in job_id.singletons():
                 num_updates.append(
                     len(self._lease_update_requests[single_job_id]))
-            updated_lease = \
-                min(num_updates) == self._jobs[job_id].scale_factor
+            updated_lease = min(num_updates) == scale_factor
+            for i, single_job_id in enumerate(job_id.singletons()):
+                self._logger.debug('{0} / {1} worker(s) for job {2} have '
+                                   'requested a lease update this '
+                                   'round'.format(
+                                       num_updates[i], scale_factor,
+                                       single_job_id))
             if not updated_lease:
                 # Job has not requested lease updates so assume it has failed.
-                self._logger.info(
+                self._logger.error(
                     'Job {0} had an extended lease but has '
                     'been unresponsive'.format(job_id))
                 self._kill_job(job_id)
                 return
             elif job_id in self._completion_events:
+                self._logger.info('Completing job {0}'.format(job_id))
+
                 # Mark job as completed.
-                self._logger.info('Completing job {0} which had an '
-                                  'extended lease'.format(job_id))
                 self._completed_jobs_in_current_round.add(job_id)
                 del self._completion_events[job_id]
 
                 # Reset metadata.
-                self._in_progress_updates[job_id] = []
-                self._lease_update_requests[job_id] = []
-                self._max_steps[job_id] = None
-            elif job_id in self._jobs_with_extended_lease:
-                # Re-dispatch jobs with extended leases that are still
-                # active and have completed early.
-                if job_id not in self._completed_jobs_in_current_round:
-                    raise RuntimeError('Job does not have lease extension '
-                                       'event but has completed!')
-                worker_ids = self._next_worker_assignments[job_id]
-                self._logger.info('Re-dispatching job {0} as it completed '
-                                  'early but has an extended lease'.format(
-                                      job_id))
-                self._try_dispatch_job(job_id, worker_ids)
+                # NOTE: We do not reset self._in_progress_updates here as
+                # multi-GPU jobs might have partially completed updates.
+                for single_job_id in job_id.singletons():
+                    self._lease_update_requests[job_id] = []
```

Comment:
Nit: I think this should either be `single_job_id` or should be out of the `for` loop